test/node/connection-emit-actions.js

```diff
@@ -0,0 +1,192 @@
+'use strict';
+
+const test = require('tap');
+const net = require('net');
+
+const jstp = require('../..');
+
+const app = require('../fixtures/application');
+
+const application = new jstp.Application(app.name, app.interfaces);
+const serverConfig = {
+  applications: [application],
+  authPolicy: app.authCallback
+};
+
+let server;
+let connection;
+
+test.beforeEach((done) => {
+  server = jstp.net.createServer(serverConfig);
+  server.listen(0, () => {
+    const port = server.address().port;
+    jstp.net.connect(app.name, null, port, 'localhost', (error, conn) => {
+      test.assertNot(error, 'must connect to server and perform handshake');
+      connection = conn;
+      done();
+    });
+  });
+});
+
+test.afterEach((done) => {
+  if (connection) {
+    connection.close();
+    connection = null;
+  }
+  server.close();
+  done();
+});
+
+test.test('must emit server and client events upon anonymous handshake',
+  (test) => {
+    const client = {
+      application: new jstp.Application('jstp', {}),
+    };
+
+    let serverConnection = null;
+    server.once('connect', (serverConn) => {
+      serverConnection = serverConn;
+    });
+
+    const port = server.address().port;
+    const socket = net.connect(port);
+    socket.on('error', () => {
+      test.fail('must create socket and connect to server');
+    });
+    socket.on('connect', () => {
+      serverConnection.on('handshakeRequest',
+        (applicationName, authStrategy, credentials) => {
+          test.equal(applicationName, app.name,
+            'application name must match');
+          test.equal(authStrategy, 'anonymous',
+            'auth strategy must be anonymous by default');
+          test.assertNot(credentials,
+            'credentials must not exist with anonymous handshake');
+        }
+      );
+
+      const transport = new jstp.net.Transport(socket);
+      const connection = new jstp.Connection(transport, null, client);
+
+      connection.on('handshake', (error, ok) => {
+        test.assertNot(error, 'handshake must not return an error');
+        test.equal(ok, app.sessionId,
+          'session id must be equal to the one provided by authCallback');
+      });
+
+      connection.handshake(app.name, null, null, (error) => {
+        test.assertNot(error, 'handshake must not return an error');
+        test.equal(connection.username, null, 'username must be null');
+        test.equal(connection.sessionId, app.sessionId,
+          'session id must be equal to the one provided by authCallback');
+        connection.close();
+        test.end();
```

Comment:
Missed this bit, is it absolute that this callback will be called after both `handshake` event handlers? If not - test will fail because of assertions after end, in that case the right way to do it is using `test.plan()`. Same applies to every test here.